Apply Nullity annotations

jchrys opened the pull request:
Motivation:

Start to use nullness annotations in Netty5 to make it easier users to understand what to expect and better ide integration.(#12527)

Modification:

Applied nullness Annotations

Result:

Better code readability & IDE integration

jchrys commented:
I just opened this PR to get reviews and comments.

jchrys commented:
Is that acceptable to place nullness annotations right before the return type?
IMHO, it is  more clearer because after all it is the return type that is notNull or nullable.

jchrys commented on common/src/main/java/io/netty5/util/concurrent/AsynchronousResult.java:
```diff
@@ -84,18 +86,18 @@
      * @throws IllegalStateException if this {@code Future} or {@link Promise} has not completed yet.
      */
-    V getNow();
+    @Nullable V getNow();
```
Is that acceptable to place nullness annotations right before the return type?
IMHO, it is more clearer because after all it is the return type that is notNull or nullable.

jchrys commented:
I tried to place nullness annotations right before the return type. Because, IMHO, it is more clearer because after all it is the return type that is nullable. Would it be acceptable to you?

hyperxpro commented:
Too many annotations look annoying when reading source code.

jchrys commented:
I agree. How about introducing a package level `@NotNull` annotation so that `@NotNull`s can be omitted?

normanmaurer commented:
@jchrys yes `@NotNull` should be default.

jchrys commented:
Thanks for your reply! 
Let me create one.

jchrys commented on license/LICENSE.jsr305.txt:
```diff
@@ -0,0 +1,28 @@
+Copyright (c) 2007-2009, JSR305 expert group
```
discussion about findbugs/jsr305 license: https://github.com/findbugsproject/findbugs/issues/128

jchrys commented on common/src/main/java/io/netty5/util/internal/NotNullByDefault.java:
```diff
@@ -0,0 +1,37 @@
+@TypeQualifierDefault({ ElementType.PARAMETER, ElementType.FIELD, ElementType.METHOD, ElementType.TYPE_USE })
+@Retention(RetentionPolicy.CLASS)
+@Target(ElementType.PACKAGE)
+@Nonnull
```
`@org.jetbrains.annotations.NotNull` couldn't be applied because of its implementation detail.

chrisvest replied:
We could open an issue with them. Then clean up later, if they add what you're missing.

jchrys replied:
Thanks for your reply. I agree. IMHO, It seems that the issue is related with IntelliJ's [impelmentation](https://github.com/JetBrains/intellij-community/blob/161e4339b5a32701cfe45bdddb9c2c03cb9d58fe/java/java-impl/src/com/intellij/codeInsight/annoPackages/Jsr305Support.java#L23) that IntelliJ only supports the nullness annotations under `javax.annotation.*` with `@TypeQualifierDefault`.

May I open an issue to ask if they will support `org.jetbrains.annotation.*` nullness annotations along with `@TypeQualifierDefault` or will have any plan to provide `@NotNullByDefault` equivalent?

+ If above annotation is acceptable, could I move on to adding the `@Nullable` annotations within the package?

chrisvest replied:
> May I open an issue to ask if they will support org.jetbrains.annotation.* nullness annotations along with @TypeQualifierDefault or will have any plan to provide @NotNullByDefault equivalent?

Yeah, please go ahead.

jchrys replied:
I've just opened it ([link](https://youtrack.jetbrains.com/issue/IDEA-303505/orgjetbrainsNotNull-does-not-work-with-javaxannotationmetaTypeQualifierDefault)).

jchrys replied:
Removed `ElementType.TYPE_USE` from `@TypeQualifierDefault` s value.
because when it was used with `@TypeQualifierDefault` it is very bug-prone.
and it effects type within the method body as well.

jchrys commented:
Hello, All.
May I request your reviews?
I just did the `common` module. (and I want to apply it module by module. Would it be acceptable?)

I tired to add `@Nullable` where it is clear. and I didn't tried to remove any alerts from static analysis tool. because I thought It would  be better in a separated commit.

jchrys commented:
I'd love to hear your feedbacks. If there are anything I could help, please let me know :D

jchrys commented:
Could I split this commit into smaller commits to get reviews and feedbacks? 

Changed to use `@org.jetbrains.NotNull` instead of `@javax.annotations.Nonnull`.
because the bug that I reported to JetBrains will be fixed in next release of ide.
and made the commit up-to-date.

release note: [link](https://youtrack.jetbrains.com/articles/IDEA-A-237338826/IntelliJ-IDEA-20223-EAP-4-22371267-build-Release-Notes)
bug report: [link](https://youtrack.jetbrains.com/issue/IDEA-303505/Constant-values-false-negative-for-orgjetbrainsNotNull-used-with-javaxannotationmetaTypeQualifierDefault)

jchrys commented:
I've force-pushed it so that it can contain minimum changes.
I would appreciate it if you could take a look.
If there are any other changes that you think would be helpful, please let me know and I'll do my best to implement them.

jchrys commented:
> Changed to use `@org.jetbrains.NotNull` instead of `@javax.annotations.Nonnull`. because the bug that I reported to JetBrains will be fixed in the next release. and made this commit up-to-date.
> 
> Could I split this commit into smaller commits to get reviews and feedbacks?
> 
> jetbrains release note: [link](https://youtrack.jetbrains.com/articles/IDEA-A-237338826/IntelliJ-IDEA-20223-EAP-4-22371267-build-Release-Notes) bug report: [link](https://youtrack.jetbrains.com/issue/IDEA-303505/Constant-values-false-negative-for-orgjetbrainsNotNull-used-with-javaxannotationmetaTypeQualifierDefault)

this issue is resolved. so I think now is good time to add `@NotNullByDefault` annotation.

jchrys commented:
Hello @normanmaurer, I believe this PR is now ready to be reviewed.
Would you kindly take a look at the changes when you have a moment?
Thank you!

jchrys commented:
This is a gentle reminder to consider applying nullity annotations in Netty5. @normanmaurer @chrisvest @franz1981
